indigo/indigo/src/main/scala/indigo/shared/formats/GenericTileset.scala

```diff
@@ -0,0 +1,39 @@
+package indigo.shared.formats
+
+import indigo._
+
+/** Utility for working with generic sprite sheets. The
+  * sheet must be uniform, with each sprite having the same
+  * width and height, and a (possibly zero-pixel) margin.
+  */
+final case class GenericTileset(
+    assetName: AssetName,
+    width: Int,
+    height: Int,
+    tileWidth: Int,
+    tileHeight: Int,
+    margin: Int
+) {
+  private val rectangles =
+    for {
+      y <- 0 until height by tileHeight + margin
+      x <- 0 until width by tileWidth + margin
+      rectangle = Rectangle(x, y, tileWidth, tileHeight)
+    } yield rectangle
```

Comment:
So this works, but you can just do this with math. You don't even need the image height:
https://github.com/PurpleKingdomGames/roguelike-starterkit/blob/701c3e2f7ffb027fae694ea900a29708ab5ffd85/roguelike-starterkit/src/main/scala/io/indigoengine/roguelike/starterkit/utils/PathFinder.scala#L160

Reply:
Of course, you may want the image height for validation, and the link above would then need to be multiplied out by the size.

The advantage of your approach is that you've pre-calculated the rectangles just one, at an instantiation cost. If someone made one of these on every frame to look up a single index, they're paying a lot for little benefit. If they make one once, build a bunch of graphics ahead of time and store them and throw this instance away, then the cost is totally worth it. However, the maths calculation is cheap, so unless I've missed something (I've yet to have my first coffee of the day so... likely!) then you might as well do the math?

Reply:
I guess this was a case of premature optimization. In my own code, I instantiate the class once, but call its apply method ~240,000 times per second, so it seemed like a no-brainer.